### DISCUSSION TITLE
Issues with copied files

### DISCUSSION THREAD

**Discussion by mrghosti3** [containers/podman#23491] [Q&A] [answered]
So I'm working on a build stage in container for my app. When buildah runs `RUN cargo build...`, cargo builds the artefacts in `/app/target/release/...`  but it can be seen only during that instruction of Dockerfile. After that all files, created after `COPY ...`, seem to disappear. Wondering if this is a bug or some misconfiguration? Any help appreciated.

Below is a fragment of Dockerfile with build stage:

```Dockerfile
FROM rust:1.80.0-alpine3.20 as builder

RUN mkdir /app

WORKDIR /app
VOLUME [ "/app" ]

COPY . /app

# RUN apk install --no-cache libcurl-dev
# RUN --mount=type=cache,target=/app/target cargo install --path .
RUN cargo build -v -r && cp target/release/app ./app
RUN ls -l
```

Podman: podman-5.1.2-1.fc40
Buildah: buildah-1.36.0-1.fc40
Distro: Fedora 40

**Comment by nalind** [comment, marked as answer]
The contents of the volume (/app) are "frozen" once it's declared with the VOLUME instruction, in that they're only modified by ADD and COPY instructions, so changes made to their contents during RUN instructions are being reverted when the instruction completes.  This is an older `docker build` behavior that we mimic and that newer versions of `docker build` stopped doing.
You can either move the VOLUME instruction to the end of the file so that changes made during RUN instructions won't be discarded when they're made, drop the VOLUME instruction entirely, or update to 5.2, where that behavior will both no longer be the default, and will be selectable with a --compat-volumes flag.

**Reply by rhatdan** [reply]
5.2 was released today, how timely.

**Reply by mrghosti3** [reply]
So I was doing "something dumb" :laughing:. Thanks for the help with this.
